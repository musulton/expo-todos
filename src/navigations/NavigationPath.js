const PATH = {
    SPLASH: "SPLASH",
    TODO_LIST: "TODO_LIST"
}

export default PATH;
