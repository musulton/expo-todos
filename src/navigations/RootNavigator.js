import {NavigationContainer} from "@react-navigation/native";
import {createNativeStackNavigator} from "@react-navigation/native-stack";

import SplashScreen from "../screens/SplashScreen/SplashScreen";
import ToDoScreen from "../screens/TodoScreen/ToDoScreen";
import LoginScreen from "../screens/LoginScreen/LoginScreen";

import PopupMenu from "../screens/TodoScreen/components/PopupMenu";
import PATH from "./NavigationPath";

const Stack = createNativeStackNavigator();

const AppNavigation = () => {
    return (
        <NavigationContainer>
            <Stack.Navigator initialRouteName={PATH.TODO_LIST}>
                <Stack.Screen name={PATH.SPLASH} component={SplashScreen} options={{headerShown: false}} />
                <Stack.Screen name={PATH.LOGIN} component={LoginScreen} options={{headerShown: false}} />
                <Stack.Group
                    screenOptions={({navigation}) => {
                        return {
                            headerStyle: {
                                backgroundColor: "#F2F2F2"
                            },
                            headerShadowVisible: false,
                            headerRight: () => <PopupMenu navigation={navigation} />
                        }
                    }}
                >
                    <Stack.Screen name={PATH.TODO_LIST} component={ToDoScreen} options={{ title: "" }} />
                </Stack.Group>
            </Stack.Navigator>
        </NavigationContainer>
    )
}

export default AppNavigation;
