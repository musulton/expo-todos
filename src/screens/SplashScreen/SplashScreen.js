import {Image, View, StyleSheet} from "react-native";

import enigma from "../../shared/assets/images/enigma.png";

const SplashScreen = () => {
    return (
        <View style={styles.container}>
            <Image source={enigma} style={styles.logo} />
        </View>
    )
}

export default SplashScreen;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: "center",
        justifyContent: "center"
    },
    logo: {
        width: 200,
        height: 200
    }
})
