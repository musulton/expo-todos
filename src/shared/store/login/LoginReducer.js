import {
    LOGIN,
    LOGOUT
} from "./LoginAction";

const initialState = {
    isLoggedIn: false
}

const LoginReducer = (state = initialState, action) => {
    switch (action.type) {
        case LOGIN: {
            return {
                isLoggedIn: true
            }
        }
        case LOGOUT: {
            return {
                isLoggedIn: false
            }
        }
        default:
            return state;
    }
}

export default LoginReducer;
