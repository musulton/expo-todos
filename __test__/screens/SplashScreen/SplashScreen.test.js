import {render, screen, waitFor} from "@testing-library/react-native";

import SplashScreen from "../../../src/screens/SplashScreen/SplashScreen";
import {onNavigate} from "../../../src/navigations/RootNavigation";
import PATH from "../../../src/navigations/NavigationPath";
import {login} from "../../../src/shared/store/login/LoginAction";

const mockGetData = jest.fn();

jest
    .mock("react-redux", () => ({
        useDispatch: jest.fn(() => jest.fn())
    }))
    .mock("../../../src/shared/utils/LocalStorage", () => ({
        __esModule: true,
        default: jest.fn(() => ({
            getData: mockGetData
        }))
    }))
    .mock("../../../src/navigations/RootNavigation", () => ({
        onNavigate: jest.fn()
    }))
    .mock("../../../src/shared/store/login/LoginAction", () => ({
        login: jest.fn()
    }))

describe("SplashScreen", () => {
    beforeAll(() => {
        jest.useFakeTimers();
    })

    beforeAll(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
    })

    test("should navigate to TODO LIST when has token", async () => {
        mockGetData.mockImplementation(() => "token123");

        render(<SplashScreen />);

        await waitFor(() => {
            jest.runOnlyPendingTimers();
        });

        expect(login).toHaveBeenCalled();
        expect(onNavigate).toHaveBeenCalledWith({
            isReplace: true,
            routeName: PATH.TODO_LIST
        })
    })

    test("should navigate to LOGIN when has not token", async () => {
        mockGetData.mockImplementation(() => null);

        render(<SplashScreen />);

        await waitFor(() => {
            jest.runOnlyPendingTimers();
        });

        expect(login).not.toHaveBeenCalled();
        expect(onNavigate).toHaveBeenCalledWith({
            isReplace: true,
            routeName: PATH.LOGIN
        })
    })

    test("should return correctly", () => {
        render(<SplashScreen />);

        expect(screen.toJSON()).toMatchSnapshot();
    })
})
