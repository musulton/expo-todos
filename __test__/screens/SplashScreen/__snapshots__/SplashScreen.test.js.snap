// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplashScreen should return correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 200,
        "width": 200,
      }
    }
  />
</View>
`;
